pkg/proxy/backend/authenticator.go

```diff
@@ -24,74 +24,123 @@ import (
+		logger.Error("require frontend capabilities", zap.Stringer("common", commonCaps), zap.Stringer("required", requiredCapabilities))
+		return errors.Wrapf(ErrCapabilityNegotiation, "require %s", requiredCapabilities&^commonCaps)
 	}
+	commonCaps := frontendCapability & proxyCapability
+	if frontendCapability^commonCaps != 0 {
+		logger.Error("frontend send capabilities unsupported by proxy", zap.Stringer("common", commonCaps), zap.Stringer("frontend", frontendCapability^commonCaps), zap.Stringer("proxy", proxyCapability^commonCaps))
+		return errors.Wrapf(ErrCapabilityNegotiation, "%s", frontendCapability^commonCaps)
+	}
+	resp := pnet.ParseHandshakeResponse(pkt)
+	auth.capability = commonCaps.Uint32()
+	auth.user = resp.User
+	auth.dbname = resp.DB
+	auth.collation = resp.Collation
+	auth.attrs = resp.Attrs
+
+	// write proxy header
+	if proxyProtocol {
+		proxy := clientIO.Proxy()
+		if proxy == nil {
+			proxy = &pnet.Proxy{
+				SrcAddress: clientIO.RemoteAddr(),
+				DstAddress: backendIO.RemoteAddr(),
+				Version:    pnet.ProxyVersion2,
+				Command:    pnet.ProxyCommandProxy,
+			}
+		}
+		if err := backendIO.WriteProxyV2(proxy); err != nil {
+			return err
+		}
+	}
+
+	// read server initial handshake
+	_, backendCapabilityU, err := auth.readInitialHandshake(backendIO)
 	if err != nil {
-		return err
+		return errors.WithStack(err)
 	}
-	if serverCapability&mysql.ClientSSL == 0 {
+	backendCapability := pnet.Capability(backendCapabilityU)
+	if common := commonCaps & backendCapability; (commonCaps^common)&^pnet.ClientSSL != 0 {
+		logger.Info("backend does not support capabilities from client&proxy", zap.Stringer("common", common), zap.Stringer("server", commonCaps^common), zap.Stringer("backend", backendCapability^common))
+	}
+	if backendCapability&pnet.ClientSSL == 0 {
 		// The error cannot be sent to the client because the client only expects an initial handshake packet.
 		// The only way is to log it and disconnect.
 		return errors.New("the TiDB server must enable TLS")
 	}
 
-	// Read the response from the client.
-	clientPkt, err := clientIO.ReadPacket()
-	if err != nil {
-		return err
-	}
-	clientCapability := binary.LittleEndian.Uint16(clientPkt[:2])
-	// A 2-bytes capability contains the ClientSSL flag, no matter ClientProtocol41 is set or not.
-	sslEnabled := uint32(clientCapability)&mysql.ClientSSL > 0
-	if sslEnabled {
-		// Upgrade TLS with the client if SSL is enabled.
-		if _, err = clientIO.UpgradeToServerTLS(frontendTLSConfig); err != nil {
-			return err
-		}
-	} else {
-		// Rewrite the packet with ClientSSL enabled because we always connect to TiDB with TLS.
-		pktWithSSL := make([]byte, len(clientPkt))
-		pnet.DumpUint16(pktWithSSL[:0], clientCapability|uint16(mysql.ClientSSL))
-		copy(pktWithSSL[2:], clientPkt[2:])
-		clientPkt = pktWithSSL
-	}
-	if err = backendIO.WritePacket(clientPkt, true); err != nil {
-		return err
+	// write SSL Packet
+	if err := backendIO.WritePacket(pkt[:32], true); err != nil {
```

Comment:
The backend must report an error because the salts don't match.

Reply:
This is a subproblem of handshaking.